### PR TITLE
Unrestricted actor dispatcher

### PR DESCRIPTION
**Pull request by cgillum**
This PR contains the initial implementation of an unrestricted actor dispatcher. It still uses the DTFx orchestration pipeline to ensure that actors are execute in a single (logical) thread, but it does not have any of the determinism requirements.

With this change, actors can safely do I/O, use bindings, execute non-durable async code, etc.

This is a WIP: Here is the status of tests that are passing (non-checked should assume to be failing):

- [x] DurableActor_ActorToAndFromBlob
- [x] DurableActor_AsyncIO (new in this PR)
- [ ] DurableActor_LargeActor
- [x] DurableActor_LockedIncrements (ExtendedSessions=False)
- [x] DurableActor_LockedIncrements (ExtendedSessions=True)
- [x] DurableActor_MultipleLockedTransfers (ExtendedSessions=False)
- [x] DurableActor_MultipleLockedTransfers (ExtendedSessions=True)
- [x] DurableActor_SignalAndCallStringStore
- [x] DurableActor_SignalThenPoll
- [x] DurableActor_SingleLockedTransfer
- [x] DurableActor_StringStoreWithCreateDelete

### CODE REVIEW

**Comment by cgillum** [comment]
At this point, the main thing I'm stuck trying to figure out is how to correctly implement locking, and many of the existing tests require this functionality.

From what I can tell, the locking code path in the actor dispatcher does **not** call `ContinueAsNew` (unlike the normal operation handling path) which means either the instance completes (and the auto-start implementation does not seem to handle this) or the code deadlocks waiting for the internal `continueAsNewSignal` task. @sebastianburckhardt I might need your help figuring out the right way of tweaking the dispatcher logic to account for the slightly different behavior of lock processing.

**Comment by sebastianburckhardt** [comment]
Great! I will take a closer look so I get a better idea.

I think the current implementation does indeed not call ContinueAsNew, but also does not complete (because as you mention, the auto-start cannot restart a terminated orchestration right now). It just sits there waiting for the next event, which may be a problem with the new dispatcher. The solution should be easy though... I think we can just call ContinueAsNew after ProcessLockRequest.

File: src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs [hunk 1]
```diff
@@ -321,6 +322,73 @@ private async Task OrchestrationMiddleware(DispatchMiddlewareContext dispatchCon
             await context.RunDeferredTasks();
         }
 
+        private async Task ActorMiddleware(DispatchMiddlewareContext dispatchContext, Func<Task> next)
+        {
+            var actorShim = dispatchContext.GetProperty<TaskOrchestration>() as TaskActorShim;
+            if (actorShim == null)
```
[the remark this answers is not in the document]
Comment (cgillum) [reply src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs]: I think you may be right. Let me look into it.

File: src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs [hunk 2]
```diff
@@ -321,6 +322,73 @@ private async Task OrchestrationMiddleware(DispatchMiddlewareContext dispatchCon
...
+                                    EventRaisedEvent eventRaisedEvent = (EventRaisedEvent)e;
+                                    actorShim.RaiseActorEvent(eventRaisedEvent.Name, eventRaisedEvent.Input);
+                                    break;
+                                default:
+                                    // TODO: Anything else?
+                                    break;
+                            }
+                        }
+
+                        if (eventsReceived && executeTask != null)
```
[the remark this answers is not in the document]
Comment (cgillum) [reply src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs]: Good point about `numberOfOperationsStarted`. I guess it's possible that we stop processing events once we hit the threshold (20), in which case we'll ContinueAsNew and then come back and process more. Do we have a test which covers this case already or does one need to be added?

I think you're right about the assertion. If `executeTask` is null, then we probably have some invalid state.

**Comment by cgillum** [comment]
I got a bit further with the CR suggestions. Basic locking scenarios seem to be working, but scenarios involving multiple locks do not seem to work when extended sessions are enabled. From what I can tell, it seems like orchestrations that fail to initially acquire a lock are never resumed even though the previous lock owner releases it.

**Comment by sebastianburckhardt** [comment]
I have been debugging a bit... not easy to repro in debugger, but the traces show that (at least with extended sessions enabled) messages are sometimes delivered twice.

**Comment by sebastianburckhardt** [comment]
Functionality-wise this looks pretty good now, but the renaming caused conflicts. I will resolve these next.

**Comment by sebastianburckhardt** [comment]
I think this is ready, except for dependency on azure/durabletask#283.

**Pull request merged**